### PR TITLE
Preserve separators in numeric literals with target=ES2021+

### LINKED ISSUE #53182: NumericLiteral text property strips the number of separators
# Bug Report

The code:

```
1_0
```

Is parsed into a NumericLiteral expression with text "10". Example on astexplorer: https://astexplorer.net/#/gist/6cb9fb64646ae587070a8a15083a46fb/4143393d63a1ce5d1e6076bd690af26b56dfa625

The ramification of this is that if you parse a document with typescript's parser and print it with typescript's printer, any numbers with separators are transformed into versions without separators.

Ideally, the text of NumericLiterals would reflect the text in the document, or if there's a separate text stored somewhere, then `ts.createPrinter` would use that more accurate version.

### 🔎 Search Terms

- NumericLiteral
- separator
- Number

### 🕗 Version & Regression Information

Exists at least from 4.7.4 to 4.9

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/IwfQDEA)

### 💻 Code

```ts
1_0
```

### 🙁 Actual behavior

Numeric separators are not included in the text of parsed numbers.

### 🙂 Expected behavior

They should be, so that parsing & re-printing can maintain fidelity.

### PR DESCRIPTION
**Pull request by danvk**
Please verify that:
* [x] There is an associated issue in the `Backlog` milestone (**required**)
* [x] Code is up-to-date with the `main` branch
* [x] You've successfully run `hereby runtests` locally
* [x] There are new or updated unit tests validating the change

Fixes #53182

The issue was that back when numeric separators were first implemented in TypeScript back in 2017 (https://github.com/microsoft/TypeScript/pull/20324#pullrequestreview-79946561) there was no ES2021 emit target (or ES2018 for that matter!) and so ESNext was used as a stand-in. When target=ES2021 was added later, the check for numeric separators wasn't updated from ESNext → ES2021.

### CODE REVIEW

**Comment by danvk** [comment]
@microsoft-github-policy-service agree

File: tests/cases/conformance/parser/ecmascript2021/numericSeparators/parser.numericSeparators.decimal.es2021.ts
[the remark this answers is not in the document]
Comment (danvk) [reply tests/cases/conformance/parser/ecmascript2021/numericSeparators/parser.numericSeparators.decimal.es2021.ts:1]: Sure, I added an `@target: es5, es2020, es2021` line to the existing conformance test. ES5 matches the old test, ES2020/21 shows that the switch happens at the appropriate version. I can remove one of those tests if three seems like overkill.

**Review by sandersn — APPROVED** [review]
This looks OK to me. @DanielRosenwasser @rbuckton is this safe to put in 5.4? Do we want it there? It's technically a fix but @RyanCavanaugh treated it as a bonus feature in the bug.

**Comment by RyanCavanaugh** [comment]
Not a feature
Not a bug
Some secret third thing?

**Comment by RyanCavanaugh** [comment]
TODO: Figure out how to avoid this kind of bug in the future? It's come up multiple times that people compare against ESNext because we don't know which ES20XX version the spec update will land in, then we forget to update the code. Maybe we should make ES2099 as an internal-only name and then rename it to e.g. ES2026 when the spec versions land?

**Comment by sandersn** [comment]
I think the right answer is a manual sweep of esnext each year as the standard is finalised. From asking @rbuckton yesterday we learned that it's usually April (provisional) or June (final).

There's no way to predict with ES20xx a particular addition will end up in, so I don't think renaming a placeholder will help. Even if each proposal gets its own placeholder, they'd still need a manual update to the year that they actually ended up in.

**Comment by danvk** [comment]
Let's all hope we're around to see the final feature set of ES2099! 😀

**Comment by RyanCavanaugh** [comment]
Sowing the grounds for the first Y2.1K bug 😭

**Pull request merged**